tests/src/FunctionalJavascript/AuthenticationTest.php

```diff
@@ -0,0 +1,409 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\Tests\apigee_edge\FunctionalJavascript;
+
+use Apigee\Edge\HttpClient\Plugin\Authentication\Oauth;
+use Drupal\apigee_edge\Plugin\KeyType\OauthKeyType;
+use Drupal\Core\Url;
+use Drupal\key\Entity\Key;
+
+/**
+ * Apigee Edge API credentials, authentication form, key integration test.
+ *
+ * @group apigee_edge
+ */
+class AuthenticationTest extends ApigeeEdgeFunctionalJavascriptTestBase {
+
+  /**
+   * Valid credentials.
+   *
+   * @var array
+   */
+  protected $validCredentials = [];
+
+  /**
+   * IDs of the created keys.
+   *
+   * @var array
+   */
+  protected $keys = [];
+
+  /**
+   * Initializes the credentials property.
+   *
+   * @return bool
+   *   True if the credentials are successfully initialized.
+   */
+  protected function initCredentials(): bool {
+    if (($username = getenv('APIGEE_EDGE_USERNAME'))) {
+      $this->validCredentials['username'] = $username;
+    }
+    if (($password = getenv('APIGEE_EDGE_PASSWORD'))) {
+      $this->validCredentials['password'] = $password;
+    }
+    if (($organization = getenv('APIGEE_EDGE_ORGANIZATION'))) {
+      $this->validCredentials['organization'] = $organization;
+    }
+    if (($endpoint = getenv('APIGEE_EDGE_ENDPOINT'))) {
+      $this->validCredentials['endpoint'] = $endpoint;
+    }
+
+    return (bool) $this->validCredentials;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    if (!$this->initCredentials()) {
+      $this->markTestSkipped('Credentials not found.');
+    }
+    parent::setUp();
+    $this->drupalLogin($this->rootUser);
+  }
+
+  /**
+   * Tests Apigee Edge key types, key providers and authentication form.
+   */
+  public function testAuthentication() {
+    // Delete pre-defined test key.
+    Key::load('test')->delete();
+    $this->assertEmpty($this->container->get('state')->get('apigee_edge.auth')['active_key']);
+    $this->assertEmpty($this->container->get('state')->get('apigee_edge.auth')['active_key_oauth_token']);
+
+    // Create a key entity using key's default type and provider, then visit
+    // authentication form and ensure that there are no available Apigee Edge
+    // keys.
+    Key::create([
+      'id' => 'default_key',
+    ])->save();
+    $this->drupalGet(Url::fromRoute('apigee_edge.settings'));
+    $this->assertSession()->pageTextContains('There is no available basic authentication key for connecting to Apigee Edge.');
+    $this->assertSession()->pageTextNotContains('There is no available OAuth key for connecting to Apigee Edge.');
```

Comment:
Why do not you define these specific message as constants in the test class at least? Copy-pasting the same message is not efficient.